### PR TITLE
add optional function arg as condition to `$if`.

### LINKED ISSUE #271: Allow passing in other types to `$.if`
Currently `$.if` only accepts a `boolean`. This means we're having to write a separate check, then force-cast, or set the value to an empty string to keep TS happy. Would be nice to pass other possibly `null`/`undefined` values to `$.if`, and then automatically get the non-null value back out in the callback.

```ts
$if<O2 extends O, T>(
    value: T | undefined | null,
    func: (qb: this, condition: T) => SelectQueryBuilder<DB, TB, O2>
  ): SelectQueryBuilder<DB, TB, MergePartial<O, O2>> {
    if (value) {
      return func(this, value)
    }
```

### PR DESCRIPTION
**Pull request by mikewuu**
Closes #271 

- [x] implement
- [x] unit test
- [x] document

### CODE REVIEW

**Comment by mikewuu** [comment]
Tried to make it throw a type error if a user tried to capture `value` when they've passed in a `boolean`, like:

```ts
 const query = ctx.db
  .selectFrom('person')
  .$if(true, (qb, val) => qb.where('first_name', '=', val))
```

Unfortunately I ran into issues trying to  use conditionals with generics, so this PR is the best I could do; `val` would be an `unknown` type.

File: test/node/src/select.test.ts [hunk 1]
```diff
@@ -814,5 +814,28 @@ for (const dialect of BUILT_IN_DIALECTS) {
       const endTime = Date.now()
       console.log((endTime - time) / N)
     })
+
+    it('should accept a conditional function for $.if', async () => {
```
Comment (igalklebanov) [inline test/node/src/select.test.ts:818]:
```suggestion
    it('should accept a conditional function for $if', async () => {
```

File: test/node/src/select.test.ts [hunk 2]
```diff
@@ -814,5 +814,28 @@ for (const dialect of BUILT_IN_DIALECTS) {
...
+      const params: {
+        name?: string
+      } = {}
+
+      params.name = 'Jennifer'
```
Comment (igalklebanov) [inline test/node/src/select.test.ts:823]: Is there a good reason for typing this const and splitting assignment to 2 steps.

```suggestion
      const params = { name: 'Jennifer' }; 
```

**Comment by igalklebanov** [comment]
Thank you! 💪

File: src/query-builder/select-query-builder.ts [hunk 3]
```diff
@@ -1678,13 +1678,33 @@ export class SelectQueryBuilder<DB, TB extends keyof DB, O>
...
+   *      (qb, name) => qb.where('first_name', '=', name)
+   *   )
+   *
+   * In the example above, `params.name` has type `string?`, but `name` has type of `string`.
    */
-  $if<O2 extends O>(
-    condition: boolean,
-    func: (qb: this) => SelectQueryBuilder<DB, TB, O2>
+  $if<O2 extends O, T>(
+    condition: boolean | (() => T | undefined),
```
[the remark this answers is not in the document]
Comment (igalklebanov) [reply src/query-builder/select-query-builder.ts:1696]: We should check for truthy, not strictly `true`, so let's update docs to reflect that.

File: test/node/src/select.test.ts (hunk 2 above)
Comment (mikewuu) [reply test/node/src/select.test.ts:823]: Only to stress that it was set after being defined, but will update to set at `const` instead:

```ts
 const params: {
        name?: string
      } = { name: 'Jennifer' }
```

**Comment by igalklebanov** [comment]
> Tried to make it throw a type error if a user tried to capture `value` when they've passed in a `boolean`, like:
> 
> ```ts
>  const query = ctx.db
>   .selectFrom('person')
>   .$if(true, (qb, val) => qb.where('first_name', '=', val))
> ```
> 
> Unfortunately I ran into issues trying to use conditionals with generics, so this PR is the best I could do; `val` would be an `unknown` type:
> 
> CleanShot 2023-01-05 at 23 18 49

Most of the time, our consumers will pass a value or function that their "truthyness" can only be determined at runtime. Consumers should be discouraged from using `$if` with `true` or `false` known at compile-time, kinda pointless no?

**Comment by mikewuu** [comment]
> Most of the time, our consumers will pass a value or function that their "truthyness" can only be determined at runtime. Consumers should be discouraged from using `$if` with `true` or `false` known at compile-time, kinda pointless no?

Agreed. Also, _always_ passing the truthy value in all cases as suggested in https://github.com/koskimas/kysely/pull/272#discussion_r1063015157 made this a non-issue as well.

File: src/query-builder/select-query-builder.ts [hunk 4]
```diff
@@ -1678,13 +1678,33 @@ export class SelectQueryBuilder<DB, TB extends keyof DB, O>
...
+    if (isBoolean(condition) && condition) {
+      return func(this, undefined as never)
+    }
+
+    if (isFunction(condition)) {
+      const value = condition()
+      if (value) {
+        return func(this, value)
+      }
     }
```
[the remark this answers is not in the document]
Comment (mikewuu) [reply src/query-builder/select-query-builder.ts]: Done 👍

File: src/query-builder/select-query-builder.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikewuu) [reply src/query-builder/select-query-builder.ts]: With it being `condition: C | (() => C)`, TS actually failed to narrow the type. That is the `value` still resolved to `string?`

I've updated it to the following instead:

```ts
condition: C | (() => C | undefined | null)
```

[the remark this answers is not in the document]
Comment (mikewuu) [reply src/query-builder/select-query-builder.ts]: Updated docs in https://github.com/koskimas/kysely/pull/272/commits/ad2ba6b5dac407cb42ab9bf30b3428519e53173b

[the remark this answers is not in the document]
Comment (igalklebanov) [reply src/query-builder/select-query-builder.ts]: > With it being condition: C | (() => C), TS actually failed to narrow the type. That is the value still resolved to string?

If consumer's condition function returns just `string` (e.g. form text input), it should be infered this way. 

```ts
const condition: () => string = () => 'blah blah blah';

$if(condition, (qb) => qb); // infers as $if<O, () => string | null> and its simply a lie.
```

You can exclude falsy values from func's argument:

```ts
type Falsy = '' | 0 | false | null | undefined;

$if<O2 extends O, C>(condition: C | (() => C), func: (qb: this, Exclude<C, Falsy>) => ...
```

[the remark this answers is not in the document]
Comment (mikewuu) [reply src/query-builder/select-query-builder.ts]: Updated in https://github.com/koskimas/kysely/pull/272/commits/e357b5403ab48f9c624440eba365a8cf73d06a9c

Thank you!! This is exactly what I was looking for. It even has the nice effect of returning `never` if a user returns a `Falsy` value:

CleanShot 2023-01-06 at 11 04 41@2x

**Comment by igalklebanov** [comment]
Didn't notice that yesterday, let's also add these changes to other builders where `.$if` exists.

**Comment by mikewuu** [comment]
@igalklebanov Done. Also updated **conditional-selects.md** recipe.

**Comment by mikewuu** [comment]
@igalklebanov @koskimas  Anything else needed for this to get merged?

**Comment by mikewuu** [comment]
@koskimas Would you like me to resolve conflicts here, or are we not too sure about this one?

**Comment by koskimas** [comment]
I don't know what to do with the `$if` method. People are overusing it, causing horrible typescript performance. I'm thinking we should maybe remove it and add a `selectIf` method. Or add `selectIf` and make this one just return the output type unconditionally which would solve the typescript performance issues.

**Comment by igalklebanov** [comment]
> I don't know what to do with the `$if` method. People are overusing it, causing horrible typescript performance. I'm thinking we should maybe remove it and add a `selectIf` method. Or add `selectIf` and make this one just return the output type unconditionally which would solve the typescript performance issues.

I like the 2nd idea. 
I'm leaning towards the utility prefix with `.$ifSelect` or `.$selectIf`, but `selectIf` is OK.
Maybe we should type-omit `.select`/`.selectAll` from `.$if`'s callback's builder.

Alternatively, we could type-pick `.select`/`.selectAll` in `.$if`'s builder, and not introduce a new method.

**Comment by koskimas** [comment]
Yep, let's figure out an alternative for `$if` and close this.

**Pull request closed without merging**